File: clients/python/ees_ingest_client.py
from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any
from urllib.request import Request, urlopen


@dataclass
class EesIngestClient:
    base_url: str
    api_key: str
    system_key: str | None = None
    system_id: str | None = None
    timeout: float = 5.0

    def _post(self, path: str, payload: dict[str, Any]) -> dict[str, Any]:
        request = Request(
            f"{self.base_url.rstrip('/')}{path}",
            data=json.dumps(payload).encode("utf-8"),
            headers={
                "Content-Type": "application/json",
                "X-EES-Ingest-Key": self.api_key,
            },
            method="POST",
        )
        with urlopen(request, timeout=self.timeout) as response:
            return json.loads(response.read().decode("utf-8"))

    def _identity(self) -> dict[str, Any]:
        result: dict[str, Any] = {}
        if self.system_key:
            result["system_key"] = self.system_key
        if self.system_id:
            result["system_id"] = self.system_id
        return result

    def telemetry(self, *, metric: str, value: Any, **fields: Any) -> dict[str, Any]:
        return self._post(
            "/api/ingest/telemetry",
            {**self._identity(), "metric": metric, "value": value, **fields},
        )

    def alert(self, *, severity: str, message: str, **fields: Any) -> dict[str, Any]:
        return self._post(
            "/api/ingest/alerts",
            {
                **self._identity(),
                "severity": severity,
                "message": message,
                **fields,
            },
        )

    def diagnostic(
        self,
        *,
        diagnostic_type: str,
        payload: dict[str, Any] | None = None,
        **fields: Any,
    ) -> dict[str, Any]:
        return self._post(
            "/api/ingest/diagnostics",
            {
                **self._identity(),
                "diagnostic_type": diagnostic_type,
                "payload": payload or {},
                **fields,
            },
        )

    def snapshot(self, *, state: dict[str, Any], **fields: Any) -> dict[str, Any]:
        return self._post(
            "/api/ingest/snapshots",
            {**self._identity(), "state": state, **fields},
        )

    def log(self, *, level: str, message: str, **fields: Any) -> dict[str, Any]:
        return self._post(
            "/api/ingest/logs",
            {
                **self._identity(),
                "level": level,
                "message": message,
                **fields,
            },
        )

    def batch(self, events: list[dict[str, Any]]) -> dict[str, Any]:
        return self._post(
            "/api/ingest/batch",
            {**self._identity(), "events": events},
        )

File: clients/typescript/eesIngestClient.ts
export type EesEventType =
  | "telemetry"
  | "alert"
  | "diagnostic"
  | "snapshot"
  | "log"
  | "ai_interaction";

export interface EesBatchEvent {
  type: EesEventType;
  data: Record<string, unknown>;
}

export interface EesIngestClientOptions {
  baseUrl: string;
  apiKey: string;
  systemKey?: string;
  systemId?: string;
}

export class EesIngestClient {
  constructor(private readonly options: EesIngestClientOptions) {}

  private identity() {
    return {
      ...(this.options.systemKey
        ? { system_key: this.options.systemKey }
        : {}),
      ...(this.options.systemId
        ? { system_id: this.options.systemId }
        : {}),
    };
  }

  private async post(path: string, body: Record<string, unknown>) {
    const response = await fetch(
      `${this.options.baseUrl.replace(/\/$/, "")}${path}`,
      {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "X-EES-Ingest-Key": this.options.apiKey,
        },
        body: JSON.stringify(body),
      },
    );

    const payload = await response.json();

    if (!response.ok) {
      throw new Error(
        payload?.detail ??
          `EES ingest returned HTTP ${response.status}.`,
      );
    }

    return payload;
  }

  telemetry(metric: string, value: unknown, fields: Record<string, unknown> = {}) {
    return this.post("/api/ingest/telemetry", {
      ...this.identity(),
      metric,
      value,
      ...fields,
    });
  }

  alert(severity: string, message: string, fields: Record<string, unknown> = {}) {
    return this.post("/api/ingest/alerts", {
      ...this.identity(),
      severity,
      message,
      ...fields,
    });
  }

  diagnostic(
    diagnosticType: string,
    payload: Record<string, unknown> = {},
    fields: Record<string, unknown> = {},
  ) {
    return this.post("/api/ingest/diagnostics", {
      ...this.identity(),
      diagnostic_type: diagnosticType,
      payload,
      ...fields,
    });
  }

  snapshot(state: Record<string, unknown>, fields: Record<string, unknown> = {}) {
    return this.post("/api/ingest/snapshots", {
      ...this.identity(),
      state,
      ...fields,
    });
  }

  log(level: string, message: string, fields: Record<string, unknown> = {}) {
    return this.post("/api/ingest/logs", {
      ...this.identity(),
      level,
      message,
      ...fields,
    });
  }

  batch(events: EesBatchEvent[]) {
    return this.post("/api/ingest/batch", {
      ...this.identity(),
      events,
    });
  }
}
